enhancements/machine-config/pre-assigning-nodes-to-machine-config-pools-4-8.md

```diff
@@ -0,0 +1,269 @@
+- Provide the ability to have changes to the CRs be reflected on existing Nodes
+- Meet critical telco use cases by shipping a solution in 4.8
+
+### Non-Goals
+
+- Establish a long-term direction for addressing this within the product.  See
+  [enhancement #717](https://github.com/openshift/enhancements/pull/717) for
+  discussion around an in-product approach
+
+## Proposal
+
+Two new CRs, a controller to watch each, and a webhook.
+
+The first CR allows the admin to define patterns and a list of labels that
+should be applied to any Node with a matching name.
+
+```go
+type LabelsSpec struct {
+	// NodeNamePatterns defines a list of node name regex patterns for which the given labels should be set.
+	// String start and end anchors (^/$) will be added automatically
+	NodeNamePatterns []string `json:"nodeNamePatterns"`
+
+	// Label defines the labels which should be set if one of the node name patterns matches
+	// Format of label must be domain/name=value
+	Labels map[string]string `json:"labels"`
+}
+```
+
+This CR is used by a MutatingAdmissionWebhook when a new Node is created to
+ensure the labels necessary for MCP placement are present from the moment the
+Node object exists.  Ensuring that the Ignition file and MCO are in agreement as
+to the correct pool assignment.
+
+A controller also watches these CRs to ensure that any additions or
+modifications are applied to existing Nodes, providing a consistent name-based
+label management experience.
+
+Deletions however, require special handling.  Deleting any label that is not
+part of a `Labels` CR would be dangerous and likely to break the cluster -
+because it cannot account for labels applied manually by the admin, or by other
+components.
+
+Therefore we define an `OwnedLabels` CRD which tells us which label domains the
+operator owns exclusively.  When a label in this domain is no longer present in
+a Labels CR, it is removed.
+
+```go
+// OwnedLabelsSpec defines the desired state of OwnedLabels
+type OwnedLabelsSpec struct {
+	// Domain defines the label domain which is owned by this operator
+	// If a node label
+	// - matches this domain AND
+	// - matches the namePattern if given AND
+	// - no label rule matches
+	// then the label will be removed
+	Domain *string `json:"domain,omitempty"`
+
+	// NamePattern defines the label name pattern which is owned by this operator
+	// If a node label
+	// - matches this name pattern AND
+	// - matches the domain if given AND
+	// - no label rule matches
+	// then the label will be removed
+	// String start and end anchors (^/$) will be added automatically
+	NamePattern *string `json:"namePattern,omitempty"`
+}
+```
+
+### User Stories
+
+1. As an admin, I want Nodes to have a specific set of labels based on their
+   https://www.clli.com driven name, so that I can target workloads to machines
+   based on hardware profile, location, etc.
+
+1. As an admin, I want Nodes to be created with a set of labels consistent with
```

Comment:
nit: 2.

Reply:
this is a formatting convention, it renders correctly. see file view: https://github.com/openshift/enhancements/blob/b0b9b0eea902adac6267b053831f22ec4eaa62ca/enhancements/machine-config/pre-assigning-nodes-to-machine-config-pools-4-8.md#user-stories

Reply:
oh ok, thanks for the heads up :+1: